Reversed the default sort order of jobs in the table view.

wxtim opened the pull request:
These changes close #1002 

Make the default sort order of the table view match the the tree view. It should respect the "Latest cycle point on top" setting in the UI settings.

I'd appreciate some help with testing this. Going to try and work it out myself as an excercise in getting used to JS & Cypress.

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [ ] Appropriate tests are included (unit and/or functional).
- [ ] Appropriate change log entry included.
- [x] No documentation update required.

AaronDCole reviewed (approved):
Short and sweet, and it works as expected!

hjoliver reviewed (approved):
:+1:  for the functionality

Regarding testing, could `tests/unit/components/cylc/table/table.vue.spec.js` be extended, and fed by data with more than one cycle point (see `tests/unit/components/cylc/table/table.data.js`)?

hjoliver reviewed (changes requested):
(un-approving, since you're going to add tests!)

AaronDCole commented:
Incase you're waiting for me to write the test here, ill do it once Ive fixed / completed the test on the resizable drawer.

hjoliver commented:
Thanks @AaronDCole - althought it looks like @wxtim was keen to try it, perhaps with advice from you. Let's see what he says...

wxtim commented:
I'm planning on spending today's development day learning about Cypress, hopefully will do this as part of that... If I haven't managed then I'll holler!

wxtim commented:
@AaronDCole I have tried and failed to produce a test and am planning on heading home soon. I might have a play during the weekend out of curiosity, but here's what I tried:

- Added data to the mocked data you pointed to, and modified the unit tests already in existence. I couldn't get my head around what the existing tests were doing though.
- Testing the column heading category in the e2e testing:
  ```javascript
    it.only('Should sort the date column by default', () => {
    cy.visit('/#/table/one')
    cy
      .get('th')
      .contains('Cycle Point')
      .parent()
      .children()
      .eq(1)
      .contains('::before')
  })
  ```
  This eventually led to me looking at https://github.com/cypress-io/cypress/issues/5860
  
- The approach I'd now try if I weren't about to leave the office would be to add extra data to the examples so that there is at least one task with a different Cycle Point then e2e test that. Is that a sane/valid approach?

wxtim commented:
@AaronDCole  - I'm hollering because the final approach risks breaking lots of other tests - @oliver-sanders  seems to think that manual testing is good enough, and has also created the related issue #1086

MetRonnie commented:
Not looked into this much, but I would suspect that unit test(s) are the way to go for this

wxtim commented:
> Not looked into this much, but I would suspect that unit test(s) are the way to go for this

Yes.

kinow reviewed (commented):
The unit tests for the tree test the sort order at the CylcTree data structure level, so maybe helpful if the table has a similar structure (can't recall if it had) https://github.com/cylc/cylc-ui/blob/21f9baac23b6c0c292c1b03a38d9c591240d0e99/tests/unit/components/cylc/tree/tree.spec.js#L46

Otherwise, I think you can modify `table.vue.spec.js`. The existing tests are all using `.length` only, I think. But you can look at the tree tests and adapt to verify if the sort is working as expected. Maybe have a few entries in the table, and then (pseudo-code, not tested):

```
const wrapper = mountFunction(...) // returns a Table component
const tableItems = wraper.findAllComponents({ name: 'TableItemName?' }) // search vue test utils selectors, e.g. https://v1.test-utils.vuejs.org/api/selectors.html#find-option-object
// either find the button/element to click to change the sort order, or start testing the default sort order
// just confirm that tableItems is sorted accordingly
```

Check out `tree.vue.spec.js`, maybe there's something useful there.

Hope that helps!

wxtim commented:
@AaronDCole  - I'm not sure whether I'm able to write a meaningful test for this - please feel free to do so, else could @hjoliver and you reconsider testing.

I have created a ticket for testing this more thoroughly using Cypress: https://github.com/cylc/cylc-ui/issues/1091

MetRonnie reviewed (changes requested):
As discussed in person, need to respect the "Latest cycle point on top" setting

AaronDCole commented on CHANGES.md:
```diff
@@ -18,6 +18,14 @@ ones in. -->
+
+[#1075](https://github.com/cylc/cylc-ui/pull/1075) - Reverse default sort order
+of the table view so it matches the tree view.
+
```
@MetRonnie are you sure you dont mean 1.3? 1.4 hasn't been released yet as far as I can tell

MetRonnie replied:
Yes, after all this enhancement won't get released until 1.4 is released

AaronDCole commented:
Ive added a unit test to ensure this sort order is being respected vs the unordered (read; reversely ordered) test data. Pending a resolution to Ronnie's query about the changes version, this should be good to go.

AaronDCole commented:
> @AaronDCole - I'm not sure whether I'm able to write a meaningful test for this - please feel free to do so, else could @hjoliver and you reconsider testing.
> 
> I have created a ticket for testing this more thoroughly using Cypress: #1091

Hey man, Ive extended the table unit test to cover this. In general I try to shy away from writing too many functional tests as I think thats often missing the point of the code your trying to test, a unit test to show that the filtered tasks are being returned in the correct order - and that they are being subsequently rendered in the correct order, should suffice.

MetRonnie commented:
> Pending a resolution to Ronnie's query about the changes version, this should be good to go.

Also, there is a setting the settings view that allows users to choose "latest cycle point on top" or not, i.e. reverse the sort order. This currently works in the tree view but not the table view.

AaronDCole commented:
> > Pending a resolution to Ronnie's query about the changes version, this should be good to go.
> 
> Also, there is a setting the settings view that allows users to choose "latest cycle point on top" or not, i.e. reverse the sort order. This currently works in the tree view but not the table view.

Ahh thank you for explaining, Ill update the MR

AaronDCole commented:
> > Pending a resolution to Ronnie's query about the changes version, this should be good to go.
> 
> Also, there is a setting the settings view that allows users to choose "latest cycle point on top" or not, i.e. reverse the sort order. This currently works in the tree view but not the table view.

Have added; In some ways I would sort of like to refactor it so that it sets the cycle point sorting (and asc preference) only if the user has that preference selected, otherwise it defaults to first field ordering, but that would need some more discussion and potentially goes agains't the point of this MR.

oliver-sanders commented:
> Have added; In some ways I would sort of like to refactor it so that it sets the cycle point sorting (and asc preference) only if the user has that preference selected, otherwise it defaults to first field ordering, but that would need some more discussion and potentially goes agains't the point of this MR.

FYI: In combination with the graph view I'm currently looking at performing sorting in a central data store.

AaronDCole commented:
Its failing because of the patch codecov, but Im not very motivated to update the test as the only change has been to based the sort order on a browser method (local storage), which has already been tested outside of this component (and as Oliver mentioned earlier we might end up changing this later anyway).

MetRonnie commented on src/components/cylc/table/Table.vue:
```diff
@@ -99,6 +99,7 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
           fluid
           class="ma-0 pa-0 w-100 h-100 left-0 top-0 position-absolute"
         >
+          {{  }}
```
@AaronDCole Why add this?

AaronDCole replied:
Ahhh no idea? I think thats a mistake, ill remove it

MetRonnie reviewed (commented):
Tested out, working as expected

hjoliver reviewed (approved):
Works for me, looks good, and Ronnie's question has been addressed... merging.
